### DISCUSSION TITLE
Force the UI to update on a model's property change trough the ViewModel

### DISCUSSION THREAD

**Discussion by Nemonek** [AvaloniaUI/Avalonia#16721] [Q&A] [answered]
Hi to everyone, I'm currently writing an app with the MVVM pattern and I have some UI elements inside a listbox which uses converters that returns a value based on a property of a model.
When I update said property I need the UI to recall the converters so that the background color can change.

Here's my situation: I have the model Spell with the IsReady property, and I then have the ViewModel SpellInfoViewModel; inside this ViewModel I have the command ReadySpell which simply changes the IsReadyProperty to true. Here come's my problem: when I edit this property I need two elements to be Updated: the first is the button I use to "ready the spell", and the second is a listbox item in a view that's loaded "under" the current one which is in a ContentControl acting as a sort of popup. Here's a photo ( Strange background colors were/are for debug purposes ):
![Immagine](https://github.com/user-attachments/assets/21c74579-68e2-4be1-adee-598bdf3e0015)
And here's some code:
``` C#
public ICommand ReadySpell { get; }
// In the SpellInfoViewModel's constructor
  ReadySpell = ReactiveCommand.Create((Spell s) =>
  {
      s.IsReady = true;
      Debug.WriteLine($"Spell {s.Name} has now the ready status set to {s.IsReady}");
  });

// Converter for the "main view" element
  public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
  {
      Spell sender = (Spell)value!;
      if (sender.IsReady)
      {
          if (sender.CastTime == CastingTime.Reaction)
              return Brushes.Orange;

          return Brushes.LightGreen;
      }

      if (sender.IsRitual)
          return Brushes.IndianRed;

      return Brushes.Gray;
  }
```

``` XAML
<!-- Code for the listbox item template. Please ignore all the other converters: those are for other functionlity of the app which do not compete this problem -->
	<ListBox.ItemTemplate>
		<DataTemplate>
			<Border Margin="0" CornerRadius="5" Padding="3" Background="{Binding, Converter={StaticResource SpellCastingTimeConv}}" BorderBrush="Black" BorderThickness="1">
				<i:Interaction.Behaviors>
					<ia:EventTriggerBehavior EventName="PointerPressed">
						<!--  
							https://github.com/AvaloniaUI/Avalonia/discussions/11962
							https://docs.avaloniaui.net/docs/basics/data/data-binding/compiled-bindings#type-casting
						-->
						<InvokeCommandAction CommandParameter="{Binding}" Command="{Binding $parent.((vm:SpellsViewModel)DataContext).OpenSpellInfoPopUp}"/>
					</ia:EventTriggerBehavior>
				</i:Interaction.Behaviors>
				<DockPanel>
					<DockPanel DockPanel.Dock="Top" VerticalAlignment="Center">
						<TextBlock FontWeight="Bold" Text="{Binding Name}"/>
						<TextBlock HorizontalAlignment="Right" Text="{Binding Level}"/>
					</DockPanel>
					<DockPanel>
						<TextBlock VerticalAlignment="Bottom" Text="{Binding CastTime}"/>
						<Border Width="50" Padding="0, 0.2, 0, 0.1" Background="{Binding, Converter={StaticResource SpellTypeConv}}" BorderThickness="1" CornerRadius="5" HorizontalAlignment="Right" VerticalAlignment="Bottom">
							<TextBlock Text="{Binding Type}" TextAlignment="Center" FontSize="10"/>
						</Border>
					</DockPanel>
				</DockPanel>
			</Border>	
		</DataTemplate>
	</ListBox.ItemTemplate>
```
The issue for the button is similar, the only difference is that said button is inside the template for items in an ItemsControl tag.

I tried looking around on google, stackoverflow and github but I cannot find what I'm looking for, so I'm asking to you.

I hope I have explained well what's my situation and what I need; if not feel free to ask for clarifications and I'll do my best to provide them.
Also, if you encounter some error o bad practices in the code I provided feel free to point them out. I'm not an expert and in fact I'm still learning both Avalonia and C#.
Thanks to everyone who will reply.

**Comment by IanRawley** [comment]
One of the main issues I see is that you're binding the spell itself to the ListBoxItem.  The Spell doesn't change, so the background colour never gets re-evaluated by the converter.  You'll need to Multibind your Background colour converter to all three of IsReady, CastingTime, and IsRitual directly assuming Spell notifies about property changes.  If Spell doesn't notify about property changes, you'll need to wrap it in something that does or find some other way to let the UI know that something has changed.

**Comment by Nemonek** [comment]
Hey, it took me a while both to understand and try. I tried using a multivalue converter and adding INotifyPropertyChanged to the Spell model. When the program reachs the multi value converter it sends to it only {{unset}} values and consequently raise an exception when I try to cast them. Looking around I found that I can apply a Fallback value, but I cannot do that in my program as I need that piece of data to be elaborated or converted, otherwise I could have just removed the converter. 
Here's the converte's code:
```C#
// I tried first with the IsReady property, then with the others but I only get {{unset}}.
  public object? Convert(IList<object?> values, Type targetType, object? parameter, CultureInfo culture)
  {
      bool sender = (bool)values[0];

      if (sender)
          return Brushes.LightGreen;
      return Brushes.Gray;
      
  }
```

And here's how I cahnged the Axaml code:
``` XAML
<ListBox.ItemTemplate>
	<DataTemplate>
		<Border Margin="0" CornerRadius="5" Padding="3"  BorderBrush="Black" BorderThickness="1">
			<Border.Background>
				<MultiBinding Converter="{StaticResource SpellCastingTimeConv}">
					<Binding Path="IsReady"/>
				</MultiBinding>
			</Border.Background>
			
			<i:Interaction.Behaviors>
				<ia:EventTriggerBehavior EventName="PointerPressed">
					<!--  
						https://github.com/AvaloniaUI/Avalonia/discussions/11962
						https://docs.avaloniaui.net/docs/basics/data/data-binding/compiled-bindings#type-casting
					-->
					<InvokeCommandAction CommandParameter="{Binding}" Command="{Binding $parent.((vm:SpellsViewModel)DataContext).OpenSpellInfoPopUp}"/>
				</ia:EventTriggerBehavior>
			</i:Interaction.Behaviors>
			<DockPanel>
				<DockPanel DockPanel.Dock="Top" VerticalAlignment="Center">
					<TextBlock FontWeight="Bold" Text="{Binding Name}"/>
					<TextBlock HorizontalAlignment="Right" Text="{Binding Level}"/>
				</DockPanel>
				<DockPanel>
					<TextBlock VerticalAlignment="Bottom" Text="{Binding CastTime}"/>
					<Border Width="50" Padding="0, 0.2, 0, 0.1" Background="{Binding, Converter={StaticResource SpellTypeConv}}" BorderThickness="1" CornerRadius="5" HorizontalAlignment="Right" VerticalAlignment="Bottom">
						<TextBlock Text="{Binding Type}" TextAlignment="Center" FontSize="10"/>
					</Border>
				</DockPanel>
			</DockPanel>
		</Border>	
	</DataTemplate>
</ListBox.ItemTemplate>
```

**Reply by IanRawley** [reply, marked as answer]
That makes sense, as Bindings typically start Unset before getting real values.  You need to type-check the inputs and respond appropriately rather than just casting.  See the example IMultiValueConverter implementation [here](https://docs.avaloniaui.net/docs/guides/data-binding/how-to-bind-multiple-properties#multibinding-example)

**Reply by Nemonek** [reply]
I looked at the link, and it provide a way of not making the application crash: isn't there a way to force bindings to send real values to the converter instead of sending Unset? Like maybe waiting to call the converter until the binding has a real value?

Edit: I mis intepreted the documentation: now that I tested it and read the docs again I got just what I needed.
Thak you very much for your help and time. I'll now mark this question as answered
